### DISCUSSION TITLE
Building avr-gdb on a RPi, which version?

### DISCUSSION THREAD

**Discussion by lkoepsel** [bloombloombloom/Bloom#109] [General]
Hi Nav,
I've decided to build all new versions of the avr tool chain on the Raspberry Pi as Debian's are quite old. I'm using 9.4 gcc, 2.39 binutils, 2.1 libc (avrdudes version) and 7.3 of avrdude.

Which version do you recommend for gdb?

Thanks!
Lief

**Comment by navnavnav** [comment]
Hey Lief,

I would recommend GDB 12.1, as that's what I use on a daily basis for AVR debugging. But a newer version should work fine.

It looks like 12.1 is available in the Debian repository for Bookworm: https://packages.debian.org/bookworm/gdb-avr. A new maintainer has come along since Sep 2022 and they seem to be quite active. The most recent build was pushed in April, for version 13.2, (for Debian unstable): https://metadata.ftp-master.debian.org/changelogs//main/g/gdb-avr/gdb-avr_13.2-4_changelog. So you may not need to build GDB from source.

**Reply by lkoepsel** [reply]
Thanks!

Funny, gcc was so old on Bookworm...5.4, that I didn't even check. :)

Lief

**Reply by navnavnav** [reply]
Yikes! 5.4 is very old.

I use GCC 13.2.0, which I built from source. I took some notes for this which I'll include below. They're just rough ramblings but they may be helpful.

```
// install this dependency first (there may be others but this is all I needed):
sudo apt-get install libiberty-dev;

and libc dependency:

sudo apt-get install libc6-dev;

-----------------------------------------------------------

We need to build binutils first...

git clone git://sourceware.org/git/binutils-gdb.git
cd binutils-gdb;
git checkout binutils-2_41-release;
mkdir ./avr-build;
cd ./avr-build;
../configure --prefix=/opt/avr-g++-13.2.0/ --target=avr --enable-languages=c,c++ --disable-nls --disable-libssp
make -j 6;
sudo make install;

------------------------------------------------------------

Then set our path:
export PATH="/opt/avr-g++-13.2.0/bin:$PATH"

------------------------------------------------------------

Now we can build GCC

git clone https://gcc.gnu.org/git/gcc.git;
cd gcc;
git checkout releases/gcc-13.2.0;
./contrib/download_prerequisites;
mkdir ./avr-g++-13.2.0-build;
cd ./avr-g++-13.2.0-build;
../configure --prefix=/opt/avr-g++-13.2.0/ --target=avr --enable-languages=c,c++ --disable-nls --disable-libssp --with-dwarf2 --enable-libstdcxx --with-avrlibc --without-headers --disable-hosted-libstdcxx;
make -j 6;
sudo make install;

--------------------------------------------------------------

Finally, build avr-libc

This didn't work. I coudln't figure out how to correctly generate the configure script. So I just downloaded a release (which contains the configure script) and built that.

cd ~/Projects/;
git clone git@github.com:avrdudes/avr-libc.git;
cd ./avr-libc/;
git checkout avr-libc-2_1_0-release;
mkdir ./build;
cd ./build;
../configure --build=`./config.guess` --host=avr --prefix=/opt/avr-g++-13.2.0/

```

The notes above were taken for a x86_64 build. Not sure if you need to do anything different for arm64

**Reply by lkoepsel** [reply]
Yeah, 5.4 is absurdly old. I did do some checking and it appears that AVR starts to diverge (deprecate?) a bit in versions newer than 11. I think (no proof), that 9.4 appears to be a sweet spot, though some on AVRFreaks along with Arduino, believe 7.4 is best.

I had no issues with building 9.4 on the RPi. It did take 2 hours and I did so on one with 8GB of RAM. I had serious issues with libc...and finally got it to work with this hack: (I had to chown to me, then chown back to root...)
```
cd ~/avr-build/avr-libc
./bootstrap
./configure --build=`./config.guess` --host=avr --prefix=/usr/local/avr
sudo chown -R lkoepsel /usr/local/avr
make -j 4
make install
sudo chown -R root /usr/local/avr
```

BTW - I'm going to document all of the steps from base Bookworm to a complete hardware debugging platform using Bloom, shortly. I'll advise when I do. It might be of value to those wanting a complete set of instructions of how to use Bloom, from a non-Linux system using the RPi as the development platform (use VS Code on PC and remote SSH into RPi).

**Comment by navnavnav** [comment]
Hello again @lkoepsel 

Just FYI: v2.0.0 has just been released. With it, comes a new build configuration: "headless builds". These builds are very similar to the build I provided you a while back, for running Bloom on your Raspberry Pi.

You can download headless builds from the [downloads page](https://bloom.oscillate.io/download).

I'm going to close this discussion now. Feel free to reopen if you wish to discuss further.
